src/goto-instrument/havoc_utils.h

```diff
@@ -0,0 +1,38 @@
+/*******************************************************************\
+
+Module: Utilities for building havoc code for expressions.
+
+Author: Saswat Padhi, saspadhi@amazon.com
+
+Date: July 2021
+
+\*******************************************************************/
+
+/// \file
+/// Utilities for building havoc code for expressions
+
+#ifndef CPROVER_GOTO_INSTRUMENT_HAVOC_UTILS_H
+#define CPROVER_GOTO_INSTRUMENT_HAVOC_UTILS_H
+
+#include <set>
+
+#include <goto-programs/goto_program.h>
+
+typedef std::set<exprt> modifiest;
+
+void append_havoc_code(
+  const source_locationt source_location,
+  const modifiest &modifies,
+  goto_programt &dest);
+
+void append_object_havoc_code_for_expr(
+  const source_locationt source_location,
+  const exprt &expr,
+  goto_programt &dest);
+
+void append_scalar_havoc_code_for_expr(
+  const source_locationt source_location,
+  const exprt &expr,
+  goto_programt &dest);
```

Comment:
Why the suffix "append"? Perhaps we can come up with more intuitive names. What do you think about `havoc_code`, `havoc_object`, and `havoc_scalar`?

Reply:
The original prefix was `build` and it was inserting goto statements at the end of the `dest` goto program. I was always confused about whether it adds the statements at the beginning or at the end, so I just changed `build` -> `append`. What part of the name is unintuitive?

The names `havoc_code`, `havoc_object`, and `havoc_scalar` seem okay but lead to the same confusion as `build`. Also notice that `append_havoc_code` takes an entire `modifies` `set`, but the `append_*_havoc_code_for_expr` functions take `exprt`.

The editor autocompletes anyway, so I just use longer function names that are more explicit about their effect.

Reply:
This is not a blocker, since it is probably just a style matter. For me, clear and concise method names yields clear code. 

> I was always confused about whether it adds the statements at the beginning or at the end, so I just changed `build` -> `append`.

I'd add this information to a header comment in each method. This will also help us to build document for this file using Doxygen. Some editors also pick up this information. For instance, take a look at this method:
https://github.com/diffblue/cbmc/blob/0bbbe195ff7e9fe55c2275dda37457de307e91a9/src/goto-instrument/contracts/contracts.h#L74-L87

> Also notice that `append_havoc_code` takes an entire `modifies` set, but the `append_*_havoc_code_for_expr` functions take `exprt`.

This is the type of information that could easily go to a comment header or inferred just by looking at the parameters.

Reply:
Thanks, I agree I should add some comments here regardless, but for now I think I will leave `append_*_havoc_code_for_expr` and `append_havoc_code`, unless there is something specific that's incorrect / confusing about these names.

> What do you think about `havoc_code`, `havoc_object`, and `havoc_scalar`?

I thought a bit more about these names, but `havoc_code` seems incorrect -- it doesn't havoc a `codet` but a `set<exprt>`.

I guess we can revisit this later when we add comments & documentation for our functions.

Reply:
Personally, I like the `append_*_` naming :-)